### PR TITLE
feat(NFDIV-680): Add awaiting payment & submitted state transitions

### PR DESCRIPTION
**Pull request by SebFlippence**
### JIRA link (if applicable) ###

[NFDIV-680](https://tools.hmcts.net/jira/browse/NFDIV-680) Payment - State Changes

### Change description ###

Transitions state between `Draft` and `AwaitingPayment` then to: `AwaitingDocuments` / `Submitted`.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenSubmitApplication.java [hunk 1]
```diff
@@ -37,29 +39,32 @@
...
 
-        CaseData caseDataCopy = details.getData().toBuilder().build();
+        CaseData data = details.getData();
+        CaseData caseDataCopy = data.toBuilder().build();
         OrderSummary orderSummary = paymentService.getOrderSummary();
         caseDataCopy.setApplicationFeeOrderSummary(orderSummary);
 
-        ListValue<Payment> paymentListValue = createPendingPayment(orderSummary.getPaymentTotal());
-        caseDataCopy.setPayments(singletonList(paymentListValue));
+        if (data.getPayments() == null || data.getPayments().isEmpty()) {
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenSubmitApplication.java:64]: I'm not sure about locking this behind this if statement. If there is a failed payment it won't execute. We might fail, go back and then want to add another payment

File: src/main/java/uk/gov/hmcts/divorce/common/model/CaseData.java [hunk 2]
```diff
@@ -735,6 +735,16 @@ public Integer getPaymentTotal() {
             .reduce(0, Integer::sum);
     }
 
+    @JsonIgnore
+    public Boolean wasLastPaymentUnsuccessful() {
+        return payments != null && payments
+            .stream()
+            .reduce((first, second) -> second)
```
Comment (linusnorton) [inline src/main/java/uk/gov/hmcts/divorce/common/model/CaseData.java:742]: it would be more useful if you named the variables id, payment

**Review by velimd — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenAddPayment.java [hunk 3]
```diff
@@ -42,24 +45,34 @@
...
             .description("Payment made")
+            .retries(120, 120)
             .grant(CREATE_READ_UPDATE, CITIZEN, CASEWORKER_DIVORCE_COURTADMIN, CASEWORKER_DIVORCE_COURTADMIN_BETA)
+            .grant(READ, CASEWORKER_DIVORCE_SUPERUSER)
             .aboutToSubmitCallback(this::aboutToSubmit);
     }
 
     public AboutToStartOrSubmitResponse<CaseData, State> aboutToSubmit(CaseDetails<CaseData, State> details,
                                                                        CaseDetails<CaseData, State> beforeDetails) {
+        log.info("Add payment about to start callback invoked");
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenAddPayment.java:59]: about to submit ?

File: src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenAddPayment.java [hunk 4]
```diff
@@ -42,24 +45,34 @@
...
 
+        log.info("Validating case data");
         List<String> submittedErrors = Submitted.validate(data);
         List<String> awaitingDocumentsErrors = AwaitingDocuments.validate(data);
         State state = details.getState();
         List<String> errors = Stream.concat(submittedErrors.stream(), awaitingDocumentsErrors.stream())
             .collect(Collectors.toList());
 
-        if (submittedErrors.isEmpty()) {
+        if (data.wasLastPaymentUnsuccessful()) {
```
Comment (nitinprabhu) [inline src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenAddPayment.java:70]: we would need test for this if block

File: src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenSubmitApplication.java (hunk 1 above)
Comment (SebFlippence) [reply src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenSubmitApplication.java:64]: This seems to be always replacing any existing payments with a mock payment, so I wasn't able to transition without checking to make sure it does replace an existing payments with a mock payment.

Comment (linusnorton) [reply src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenSubmitApplication.java:64]: I'm working on creating you a real one but I'm starting to second guess myself as to whether it should be done in the frontend

File: src/main/java/uk/gov/hmcts/divorce/common/model/CaseData.java (hunk 2 above)
Comment (SebFlippence) [reply src/main/java/uk/gov/hmcts/divorce/common/model/CaseData.java:742]: Technically this is the current and previous item in the stream, I'll update this to reflect 👍

File: src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenAddPayment.java (hunk 4 above)
Comment (SebFlippence) [reply src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenAddPayment.java:70]: 👍 Fixed in ceb4fe7

File: src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenAddPayment.java (hunk 3 above)
Comment (SebFlippence) [reply src/main/java/uk/gov/hmcts/divorce/citizen/event/CitizenAddPayment.java]: 👍 Good spot, fixed in ceb4fe7

**Pull request merged**